### DISCUSSION TITLE
"Light bleed compensation" tool performing task backwards?

### DISCUSSION THREAD

**Discussion by bryanseger** [sn4k3/UVtools#868] [General]
I'm running into a situation where dimming pixels on a bridge or overhang area may be the solution to minimize issues with light bleed. I have attached the .STL and .ctb in this ZIP file for an example. 
[light bleed comp demo files.zip](https://github.com/sn4k3/UVtools/files/15002499/light.bleed.comp.demo.files.zip)

Manually dimming the pixels in UVtools (with Pixel Editor) seems to be working, but it is painfully slow. I started looking through the available tools and found "light bleed compensation". I believe this is exactly what I want, but it's performing its function opposite to what I was expecting. It grays out the pixels that I'd expect to be white, and it whitens the pixels that I want gray.

I tried all 3 pixel lookup modes, but all resulted in a similar fashion. I even tried inputting the "dim subsequent pixels by:" values in opposite order, but that did not help either.

Am I missing something here? Using this tool makes the overhanging pixels brighter, which'll result in more light bleed, not less. Thanks for the help!

**Comment by sn4k3** [comment]
That tool compares below pixels, if they are repeated below it will dim the current ones.
The tool prevent over cure layers with transparent resins where light pass from current layer to the previous
If you have a overhang with no pixels below it is expected to full cure the pixels.

In other words the tool is made for Z bleed and not XY bleed

**Comment by bryanseger** [comment]
I think I'm looking for the opposite of that then. I too am concerned with Z bleed. I'm not worried about X/Y.

I have an overhang, with no pixels directly below it, but I want that first overhang layer to have, say, 25% brightness, and then the next layer to have 50% brightness, and then the next at 75%, and then full brightness after that. Something like below. (and I used Pixel Editor and Paint to make this image, but this is just illustrative...I'm not actually using Pixel Editor in this fashion).

![image](https://github.com/sn4k3/UVtools/assets/4185392/ea2bcd8a-045f-4c30-9c91-17e771ad7f2a)


Hopefully the above makes sense. I start the first 'bridge' layer and next few layers at a lower brightness to limit how much Z bleed occurs inside the black area. Is something like this achievable with UVtools?

**Reply by sn4k3** [reply]
It doesn't make sense to me. If you have an overhang (bridge) you want to exposure that at full time and power, I'm not seeing a 25% overhang to form correctly (You will loss that detail). However if just one row or tiny area as you show the XY bleed will help. Still I can't understand your sample. It looks more as an AntiAlias effect than a bleed compensation.

**Reply by bryanseger** [reply]
I have done additional testing to prove this theory. Attached is the modified CTB that I ran yesterday, plus the data to support that dimming the pixels aids in less Z bleed. 

In the CTB, there are 4 pieces being printed and are all identical with the exception of pixel editing. One is "as-is" (full brightness), and the other 3 have varying amounts of pixel dimming (if you go to the very last few layers, you'll see where I have physically labeled the parts so I can tell them apart after they've been printed). All pixel dimming was done via the Pixel Editor, which works great but can be time consuming. You'll see on layers 61 and 101, I begin 10 layers of dimmed pixels (these layers are where bridges and overhangs begin). (You'll also see similar dimming occur beginning at layer 141, but I did not measure these portions today, so I have not included them in the data.)

To keep my time to a minimum in manually editing and dimming pixels, I went for an easy number of 10 layers at reduced brightness levels of 75%, 50%, and 25%. There's definitely a more scientific approach towards this, but just for confirming my suspicions, I kept it simple for now. Pixel dimming was not applied on the 3 widest bridges just in the interest of time. (also, after all the pixel dimming was performed, I then applied a mask to the entire print...so technically every pixel is dimmed, but the 25/50/75% areas are still substantially more dim than the normal masked areas)

I ran the CTB and then measured the parts on a Micro-Vu Sol machine. The parts were backlit, so I only saw (and measured) the 'worst' of the Z bleed.
![image](https://github.com/sn4k3/UVtools/assets/4185392/a4882e7f-d27b-47d5-aa4e-f2afe96263c6)


You can see in the data that all overhangs and bridges (with the exception of the 3 widest ones that I didn't dim) that having 10 layers of dimmed pixels at the beginning of these overhangs/bridges results in less Z bleed. The ideal/perfect dimension for the bridges is zero, and the ideal/perfect dimension for the overhangs is 1.000mm, and the 25% brightness parts had the closest dimensions to these ideal/perfect numbers. 

![image](https://github.com/sn4k3/UVtools/assets/4185392/d78d2dc9-b0e1-43bb-9aa5-3cdd223dd368)

![image](https://github.com/sn4k3/UVtools/assets/4185392/e0f7c01d-61d1-4a7f-a486-633e0569ae91)


Having this data now, I think it's safe to say there's a correlation here to where dimming the pixels at the beginning of an overhang/bridge can result in less Z bleed. I'd like to continue testing this, but manually editing/dimming is very time consuming. I'm wondering if there can be an option added to the "light bleed compensation" tool (or a new tool altogether) that mimics this behavior in a controllable fashion?
[041724 example files.zip](https://github.com/sn4k3/UVtools/files/15016975/041724.example.files.zip)

**Reply by sn4k3** [reply]
I've modified the tool with the inverse, check if this is the desired result. (Note that I've set the model to white and then apply the tool to whole model with 4 dim steps)

[bridge and cantilever test 041624 masked _02h32m_2ml_MODIFIED.zip](https://github.com/sn4k3/UVtools/files/15017594/bridge.and.cantilever.test.041624.masked._02h32m_2ml_MODIFIED.zip)

**Reply by bryanseger** [reply]
Yes, that's very close to what I'm looking for!

How did you accomplish that? I see a lot of interesting options in the "pixel arithmetic" tool, but I haven't used any of those yet. Or, did you accomplish this on your end with something custom?

Ideally, how you had 4 dim steps in your example file, it would be nice to be able to control how many dim steps there are and the levels of brightness. Something like, if I type in "51,76,102,127,153,178,204,229", then it would dim 8 layers at those increasing brightness levels following any layer of black pixels.

**Reply by sn4k3** [reply]
![image](https://github.com/sn4k3/UVtools/assets/113281/c9ec4eeb-f200-4d9a-9659-4f539bdb288d)
By modified code. Currently no tool can achieve this.

**Reply by bryanseger** [reply]
Very nice. Is that a code you can send me, and/or is that being implemented in the next update?

**Reply by sn4k3** [reply]
Will be on the next update.
If you don't want to wait I can send you the line to change, if you able to compile your own UVtools.

**Reply by bryanseger** [reply]
I currently do not know how to compile my own UVtools, but if it's easy to list off a few lines of instructions for me, then I can give it a try!

I was about to set up another test build manually, but if you're able to send the line to change and are able to walk me though compiling, I would love to test this out today.

**Reply by sn4k3** [reply]
You need the dotnet and C# workloads.
[Visual Studio Community](https://visualstudio.microsoft.com/vs/community/) for example will help you choose and install the workloads. (Checkbox on .NET Desktop development)
![image](https://github.com/sn4k3/UVtools/assets/113281/c047bf75-41f7-49c6-8fbc-0f85cef07e47)
Open UVtools.sln with it and run (F5), it will run on debug by default.

It worth the try as you can also learn scripting, which will allow you to do your own code and run in UVtools release: https://github.com/sn4k3/UVtools/tree/master/Scripts/UVtools.ScriptSample

Still the release should be up in a few hours

**Reply by sn4k3** [reply]
You can now test in the new release

**Reply by bryanseger** [reply]
Heck yea, thanks Tiago! I have a test print running right now. I'll see how it looks Monday morning.

One thing I did notice...whenever I go to "light bleed compensation" tool now, it automatically jumps to layer 30 for some reason, even if I'm at layer 0 when I go to start the tool. Any idea why that is?

**Reply by sn4k3** [reply]
It jumps to normal layers. Skipping bottoms. You can override default profile.
